src/bookshelf/src/bksw_sec.erl

```diff
@@ -21,123 +22,405 @@
+% maybe early bails with throws are harder to debug?), but it avoids
+% throws if that is desirable.
+%
+% Or heck, maybe structuring case statements to eliminate throws wouldn't be as bad
+% as I am assuming. Or maybe I'm missing something.
+%
+% Anyway, I'm flexible.  For now, assuming throws until we get things sorted out.
+
+auth(RequestId, Req0, #context{reqid = ReqId} = Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString, Headers0, VerificationType) ->
     try
-        mini_s3:make_signed_url_authorization(SecretKey,
-                                              erlang:list_to_existing_atom(Method),
-                                              Path,
-                                              Expires,
-                                              Headers)
+        % CODE REVIEW: Host is used in the generation of Config which is used in both verification types.
+
+        case Host = wrq:get_req_header("Host", Req0) of
+            undefined -> throw({RequestId, Req0, Context});
+            _         -> ok
+        end,
+
+        % CODE REVIEW: Used in obtaining CredentialScopeDate, which is used for Date, which is used in both verification types
+
+        [AWSAccessKeyId, CredentialScopeDate, Region | _] =
+            case parse_x_amz_credential(Credential) of
+                {error,      _} -> throw({RequestId, Req0, Context});
+                {ok, ParseCred} -> ParseCred
+            end,
+
+        % CODE REVIEW: Date is used in both verification types.
+
+        % https://docs.aws.amazon.com/general/latest/gr/sigv4-date-handling.html
+        DateIfUndefined = wrq:get_req_header("date", Req0),
+        case {_,  Date} = get_check_date(XAmzDate, DateIfUndefined, CredentialScopeDate) of
+            {error,  _} -> throw({RequestId, Req0, Context});
+            {ok,     _} -> ok
+        end,
+
+        % CODE REVIEW: Used in generating Config which is used in both verification types
+        AccessKey = bksw_conf:access_key_id(    Context),
+        SecretKey = bksw_conf:secret_access_key(Context),
+
+        % CODE REVIEW: used in both verification types
+        Headers = process_headers(Headers0),
+
+        % CODE REVIEW: used in both verification types
+        SignedHeaderKeys = parse_x_amz_signed_headers(SignedHeaderKeysString),
+        SignedHeaders = get_signed_headers(SignedHeaderKeys, Headers, []),
+
+        % CODE REVIEW: used in both verification types
+        RawMethod = wrq:method(Req0),
+        Method = list_to_atom(string:to_lower(erlang:atom_to_list(RawMethod))),
+
+        % CODE REVIEW: used in both verification types
+        Path = wrq:path(Req0),
+
+        % CODE REVIEW: used in both verification types
+        Config = mini_s3:new(AccessKey, SecretKey, Host),
+
+        % CODE REVIEW: AltSignedHeaders used in both verification types
+
+        % replace host header with alternate host header
+        AltHost = mini_s3:get_host_toggleport(Host, Config),
```

Comment:
Q

Reply:
This code was created before we made the change to the client side to always send host:port.  I can try to remove the alt sig calculation and see what happens.

Reply:
The function could live anywhere, but at the time of writing this I thought the most logical place for it was mini_s3.  The function was used by other modules at the time, so mini_s3 seemed like a good 'utility' location.  But the code has been through many rewrites and refactors, so I'm not sure if it's still used in other places.  I'll check.

Reply:
Per code review with @markan and @PrajaktaPurohit on 12/15/2020 it was decided to keep the alt signature calculation for the time being.